gpjax/dataset.py

```diff
@@ -15,30 +15,55 @@
 
 from dataclasses import dataclass
 
-from beartype.typing import Optional
+from beartype.typing import (
+    Optional,
+    Union,
+)
 import jax.numpy as jnp
-from jaxtyping import Num
+from jaxtyping import (
+    Bool,
+    Num,
+)
 from simple_pytree import Pytree
 
 from gpjax.typing import Array
 
 
+class _Missing:
+    """Sentinel class for not-yet-computed mask"""
```

Comment:
Could you expand a little on this description please? To me, it doesn't fully explain the class' purpose.

Also, what is a sentinel class?

Reply:
On reflection, why do we even need this class?

Reply:
The `Dataset` class holds either a boolean array or `None`, however at initialization `jnp.isnan` will be computed every time to ensure that there are no missing values in the data. This can be a lot of overhead especially when constructing batches (see `get_batch` in `fit.py`): if the original Dataset has `mask=None` there is no reason to do the computation for the subset. So the logic here is:
- if nothing is specified the default is this sentinel class, we compute the mask if needed
- if `None` is passed explicitly, we skip the computation and the mask is just `None`

In short, we need this class to provide a clear differentiation between an intentionally provided value of `None` and a default "unset" value for the mask attribute. 

In these situations the "sentinel object/value" pattern is useful, see https://python-patterns.guide/python/sentinel-object/

Reply:
This looks better to me than the solution I've always used (introduce an additional boolean variable `use_mask`). Maybe yet another would be to have a "Dataset" and a "MaskedDataset" class. I would be ok with the sentinel.

Reply:
another solution would be to use string constants, which is what some python libs do and you don't have to introduce a new class for it.

```
@dataclass
class Dataset:
    mask: Optional[Union[Bool[Array, "N Q"], str]] = "No Mask"
```

one hint: `Optional[...]` is a shorthand notation for `Union[..., None]` so your original type hint (` mask: Optional[Union[Bool[Array, "N Q"], None]] = _Missing()`) is doubling things.